In [2]:
import os
import sys
from dotenv import load_dotenv,find_dotenv

os.path.abspath(os.path.join('../Scripts'))


In [5]:
from Scripts.ranking import evaluate_prompt

ModuleNotFoundError: No module named 'Scripts'

In [5]:
load_dotenv(find_dotenv())

True

In [96]:
file_path = '../test-dataset/test-data.json'
# Open the file in read mode
with open(file_path, 'r') as file:
    # Read the contents of the file
    file_contents = file.read()
file_contents 

'[\n    {\n        "prompt": "What is the purpose of Llama 2?",\n        "ground_truth": "Llama 2 is a collection of pretrained and fine-tuned large language models optimized for dialogue use cases."\n    },\n    {\n        "prompt": "How do Llama 2 models perform compared to open-source chat models?",\n        "ground_truth": "Llama 2 models generally outperform open-source chat models on most benchmarks."\n    },\n    {\n        "prompt": "Are Llama 2 models a suitable substitute for closed-source models?",\n        "ground_truth": "Based on humane evaluations for helpfulness and safety, Llama 2 models may be a suitable substitute for closed-source models."\n    },\n    {\n        "prompt": "What is the approach to fine-tuning and safety in Llama 2?",\n        "ground_truth": "A detailed description of the approach to fine-tuning and safety in Llama 2 is provided, similar to ChatGPT, BARD, and Claude."\n    },\n    {\n        "prompt": "What limits progress in AI alignment research w

In [122]:
file_path = '../prompts/context.txt'
# Open the file in read mode
with open(file_path, 'r') as file:
    # Read the contents of the file
    context = file.read()
context

'Alan Schelten Ruan Silva Eric Michael Smith Ranjan Subramanian Xiaoqing Ellen Tan Binh Tang\nRoss Taylor Adina Williams Jian Xiang Kuan Puxin Xu Zheng Yan Iliyan Zarov Yuchen Zhang\nAngela Fan Melanie Kambadur Sharan Narang Aurelien Rodriguez Robert Stojnic\nSergey Edunov Thomas Scialom\x03\nGenAI, Meta\nAbstract\nIn this work, we develop and release Llama 2, a collection of pretrained and fine-tuned\nlarge language models (LLMs) ranging in scale from 7 billion to 70 billion parameters.\nOur fine-tuned LLMs, called L/l.sc/a.sc/m.sc/a.sc /two.taboldstyle-C/h.sc/a.sc/t.sc , are optimized for dialogue use cases. Our\nmodels outperform open-source chat models on most benchmarks we tested, and based on\nourhumanevaluationsforhelpfulnessandsafety,maybeasuitablesubstituteforclosedsource models. We provide a detailed description of our approach to fine-tuning and safety\nasChatGPT,BARD,andClaude. TheseclosedproductLLMsareheavilyfine-tunedtoalignwithhuman\npreferences, which greatly enhances their

In [101]:
import json

# file_contents = file_contents

# Load the JSON data
data = json.loads(file_contents)

# Separate prompts and ground truths
questions = [item['prompt'] for item in data]
ground_truths = [item['ground_truth'] for item in data]



In [102]:
print("\nGround Truths:")
for ground_truth in ground_truths:
    print(ground_truth)



Ground Truths:
Llama 2 is a collection of pretrained and fine-tuned large language models optimized for dialogue use cases.
Llama 2 models generally outperform open-source chat models on most benchmarks.
Based on humane evaluations for helpfulness and safety, Llama 2 models may be a suitable substitute for closed-source models.
A detailed description of the approach to fine-tuning and safety in Llama 2 is provided, similar to ChatGPT, BARD, and Claude.
The lack of transparency and reproducibility in the fine-tuning process of closed-source models limits progress in AI alignment research within the community.


In [100]:
print("\nquestions:")
for question in questions:
    print(question)



questions:
What is the purpose of Llama 2?
How do Llama 2 models perform compared to open-source chat models?
Are Llama 2 models a suitable substitute for closed-source models?
What is the approach to fine-tuning and safety in Llama 2?
What limits progress in AI alignment research within the community?


In [78]:
main_prompt = "what is special about lama2?"
test_cases = questions
# result = evaluate_prompt(main_prompt, test_cases)
# print(result)
test_cases

['What is the purpose of Llama 2?',
 'How do Llama 2 models perform compared to open-source chat models?',
 'Are Llama 2 models a suitable substitute for closed-source models?',
 'What is the approach to fine-tuning and safety in Llama 2?',
 'What limits progress in AI alignment research within the community?']

In [106]:
result = evaluate_prompt(main_prompt, test_cases)
print(result)

{'main_prompt': {'Monte Carlo Evaluation': 2.05, 'Elo Rating Evaluation': 1504.2019499940866}, 'test_case_1': {'Monte Carlo Evaluation': 2.0, 'Elo Rating Evaluation': 1504.2019499940866}, 'test_case_2': {'Monte Carlo Evaluation': 1.99, 'Elo Rating Evaluation': 1519.2019499940866}, 'test_case_3': {'Monte Carlo Evaluation': 2.14, 'Elo Rating Evaluation': 1504.2019499940866}, 'test_case_4': {'Monte Carlo Evaluation': 1.88, 'Elo Rating Evaluation': 1519.2019499940866}, 'test_case_5': {'Monte Carlo Evaluation': 2.04, 'Elo Rating Evaluation': 1504.2019499940866}}


{'main_prompt': {'Monte Carlo Evaluation': 2.08, 'Elo Rating Evaluation': 1519.2019499940866}, 'test_case_1': {'Monte Carlo Evaluation': 2.05, 'Elo Rating Evaluation': 1504.2019499940866}, 'test_case_2': {'Monte Carlo Evaluation': 2.05, 'Elo Rating Evaluation': 1519.2019499940866}, 'test_case_3': {'Monte Carlo Evaluation': 1.97, 'Elo Rating Evaluation': 1519.2019499940866}, 'test_case_4': {'Monte Carlo Evaluation': 2.1, 'Elo Rating Evaluation': 1489.2019499940866}, 'test_case_5': {'Monte Carlo Evaluation': 1.99, 'Elo Rating Evaluation': 1489.2019499940866}}


## RAGAS Evaluation

In [91]:
import requests
from langchain.document_loaders import TextLoader
from langchain.text_splitter import CharacterTextSplitter  
from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores import Weaviate
# import weaviate
# from weaviate.embedded import EmbeddedOptions
from dotenv import load_dotenv,find_dotenv
from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores import Weaviate
# import weaviate
# from weaviate.embedded import EmbeddedOptions
from dotenv import load_dotenv,find_dotenv
# 
from langchain.chat_models import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain.schema.runnable import RunnablePassthrough
from langchain.schema.output_parser import StrOutputParser

In [154]:
questions_list=questions[:3] 
questions_list

['What is the purpose of Llama 2?',
 'How do Llama 2 models perform compared to open-source chat models?',
 'Are Llama 2 models a suitable substitute for closed-source models?']

In [155]:
ground_truths_lists = [[item] for item in ground_truths[:3]]

# Print the modified list of lists
print("Ground Truths (List of Lists):")
for sublist in ground_truths_lists:
    print(sublist)

Ground Truths (List of Lists):
['Llama 2 is a collection of pretrained and fine-tuned large language models optimized for dialogue use cases.']
['Llama 2 models generally outperform open-source chat models on most benchmarks.']
['Based on humane evaluations for helpfulness and safety, Llama 2 models may be a suitable substitute for closed-source models.']


In [156]:
context_lists = [
    ["Alan Schelten Ruan Silva Eric Michael Smith Ranjan Subramanian Xiaoqing Ellen Tan Binh Tang",
    "Ross Taylor Adina Williams Jian Xiang Kuan Puxin Xu Zheng Yan Iliyan Zarov Yuchen Zhang",
    "Angela Fan Melanie Kambadur Sharan Narang Aurelien Rodriguez Robert Stojnic",
    "Sergey Edunov Thomas Scialom\x03",
    "GenAI, Meta",
    "Abstract"],
    
    ["In this work, we develop and release Llama 2, a collection of pretrained and fine-tuned",
    "large language models (LLMs) ranging in scale from 7 billion to 70 billion parameters.",
    "Our fine-tuned LLMs, called L/l.sc/a.sc/m.sc/a.sc /two.taboldstyle-C/h.sc/a.sc/t.sc , are optimized for dialogue use cases. Our",
    "models outperform open-source chat models on most benchmarks we tested, and based on",
    "ourhumanevaluationsforhelpfulnessandsafety,maybeasuitablesubstituteforclosedsource models. We provide a detailed description of our approach to fine-tuning and safety",
    "asChatGPT,BARD,andClaude. TheseclosedproductLLMsareheavilyfine-tunedtoalignwithhuman",
    "preferences, which greatly enhances their usability and safety. This step can require significant costs in",
    "computeandhumanannotation,andisoftennottransparentoreasilyreproducible,limitingprogresswithin",
    "the community to advance AI alignment research."],

    ["In this work, we develop and release Llama 2, a family of pretrained and fine-tuned LLMs, L/l.sc/a.sc/m.sc/a.sc /two.taboldstyle and",
    "L/l.sc/a.sc/m.sc/a.sc /two.taboldstyle-C/h.sc/a.sc/t.sc , at scales up to 70B parameters. On the series of helpfulness and safety benchmarks we tested,",
    "L/l.sc/a.sc/m.sc/a.sc /two.taboldstyle-C/h.sc/a.sc/t.sc models generally perform better than existing open-source models. They also appear to",
    "be on par with some of the closed-source models, at least on the human evaluations we performed"]
]

# Print the list of lists
for sublist in context_lists:
    print(sublist)


['Alan Schelten Ruan Silva Eric Michael Smith Ranjan Subramanian Xiaoqing Ellen Tan Binh Tang', 'Ross Taylor Adina Williams Jian Xiang Kuan Puxin Xu Zheng Yan Iliyan Zarov Yuchen Zhang', 'Angela Fan Melanie Kambadur Sharan Narang Aurelien Rodriguez Robert Stojnic', 'Sergey Edunov Thomas Scialom\x03', 'GenAI, Meta', 'Abstract']
['In this work, we develop and release Llama 2, a collection of pretrained and fine-tuned', 'large language models (LLMs) ranging in scale from 7 billion to 70 billion parameters.', 'Our fine-tuned LLMs, called L/l.sc/a.sc/m.sc/a.sc /two.taboldstyle-C/h.sc/a.sc/t.sc , are optimized for dialogue use cases. Our', 'models outperform open-source chat models on most benchmarks we tested, and based on', 'ourhumanevaluationsforhelpfulnessandsafety,maybeasuitablesubstituteforclosedsource models. We provide a detailed description of our approach to fine-tuning and safety', 'asChatGPT,BARD,andClaude. TheseclosedproductLLMsareheavilyfine-tunedtoalignwithhuman', 'preferences, wh

In [160]:
import pickle

# Load Answer list from the file
with open('answer_list.pkl', 'rb') as file:
    Answer = pickle.load(file)

# Now, Answer contains the list of responses from the first notebook
print(Answer)


['The purpose of Llama 2 is to provide a collection of pretrained and fine-tuned large language models optimized for dialogue use cases, outperforming open-source chat models and potentially serving as a substitute for closed-source models.', 'Llama 2 models generally perform better than existing open-source chat models.', 'Yes, Llama 2 models generally perform better than existing open-source models and appear to be on par with some closed-source models based on human evaluations.']


In [161]:
from datasets import Dataset
data = {
    "question": questions_list, # list 
    "answer": Answer, # list
    "contexts": context_lists, # list list
    "ground_truths": ground_truths_lists # list Lists
}

# Convert dict to dataset
dataset = Dataset.from_dict(data)

In [163]:
from ragas import evaluate
from ragas.metrics import (
    faithfulness,
    answer_relevancy,
    context_recall,
    context_precision,
)

result = evaluate(
    dataset = dataset, 
    metrics=[
        context_precision,
        context_recall,
        faithfulness,
        answer_relevancy,
    ],
)

df = result.to_pandas()

evaluating with [context_precision]


100%|██████████| 1/1 [00:06<00:00,  6.86s/it]


evaluating with [context_recall]


100%|██████████| 1/1 [00:01<00:00,  1.78s/it]


evaluating with [faithfulness]


100%|██████████| 1/1 [00:04<00:00,  4.33s/it]


evaluating with [answer_relevancy]


100%|██████████| 1/1 [00:04<00:00,  4.30s/it]


In [164]:
df

question  \
0                    What is the purpose of Llama 2?   
1  How do Llama 2 models perform compared to open...   
2  Are Llama 2 models a suitable substitute for c...   

                                              answer  \
0  The purpose of Llama 2 is to provide a collect...   
1  Llama 2 models generally perform better than e...   
2  Yes, Llama 2 models generally perform better t...   

                                            contexts  \
0  [Alan Schelten Ruan Silva Eric Michael Smith R...   
1  [In this work, we develop and release Llama 2,...   
2  [In this work, we develop and release Llama 2,...   

                                       ground_truths  context_precision  \
0  [Llama 2 is a collection of pretrained and fin...               0.00   
1  [Llama 2 models generally outperform open-sour...               0.25   
2  [Based on humane evaluations for helpfulness a...               0.25   

   context_recall  faithfulness  answer_relevancy  
0             1.0           0.0          1.000000  
1             1.0           1.0          0.979799  
2             1.0           1.0          0.933559

#### Integration with Retrieval-Augmented Generation Assessment:
##### Monte Carlo for Robustness Testing: Use Monte Carlo simulations to test the robustness of the RAG system across a wide range of possible retrieval scenarios. This helps in understanding how different types of retrieved information can impact the quality of the generated content.
##### Elo Rating for Continuous Improvement: Utilize the Elo rating system to continuously assess and improve the RAG model. By comparing new outputs with previous ones and adjusting ratings accordingly, the system can learn which types of retrieval-augmented generations work best.